pkg/server/cloud/gce/instances.go

```diff
@@ -0,0 +1,445 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/elotl/kip/pkg/api"
+	"github.com/elotl/kip/pkg/server/cloud"
+	"github.com/elotl/kip/pkg/util"
+	"k8s.io/klog"
+
+	"google.golang.org/api/compute/v1"
+)
+
+func convertLabelKeys(labels map[string]string) map[string]string {
+	convertedLabels := make(map[string]string)
+	for k, v := range labels {
+		switch k {
+		case cloud.ControllerTagKey:
+			k = "kip-controller-id"
+		case cloud.NametagTagKey:
+			k = "kip-nametag"
+		case cloud.NamespaceTagKey:
+			k = "kip-namespace"
+		case cloud.PodNameTagKey:
+			k = "kip-pod-name"
+		default:
+			k = strings.ToLower(k)
+		}
+		v = strings.ToLower(v)
+		v = replaceReservedLabelChars(v)
+		convertedLabels[k] = v
+	}
+	return convertedLabels
+}
+
+func (c *gceClient) getInstanceSpec(instanceID string) (*compute.Instance, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	instance, err := c.service.Instances.Get(c.projectID, c.zone, instanceID).Context(ctx).Do()
+	if err != nil {
+		klog.Errorf("error retrieving instance specification: %v", err)
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (c *gceClient) getInstanceStatus(instanceID string) (string, error) {
+	instance, err := c.getInstanceSpec(instanceID)
+	if err != nil {
+		klog.Errorf("error retrieving instance status: %v", err)
+		return "", err
+	}
+	return instance.Status, nil
+}
+
+func (c *gceClient) getInstanceLabels(nodeName string) map[string]string {
+	nametag := c.createUnboundNodeNameTag()
+	return map[string]string{
+		"name":             nametag,
+		"node":             nodeName,
+		controllerLabelKey: c.controllerID,
+		nametagLabelKey:    c.nametag,
+	}
+}
+
+func (c *gceClient) getAttachedDiskSpec(isBoot bool, size int64, name, typeURL, imageURL string) []*compute.AttachedDisk {
+	if size < minimumDiskSize {
+		klog.V(2).Infof("GCE does not allow disk smaller than 10GiB. requested size: %dGiB, using default: 10GiB", size)
+		size = minimumDiskSize
+	}
+	diskSpec := []*compute.AttachedDisk{
+		{
+			Boot:       isBoot,
+			AutoDelete: true,
+			// RW or RO
+			Mode: "READ_WRITE",
+			// SCSI or NVMe
+			Interface: "SCSI",
+			InitializeParams: &compute.AttachedDiskInitializeParams{
+				DiskName:    name,
+				DiskSizeGb:  size,
+				DiskType:    typeURL,
+				SourceImage: imageURL,
+			},
+		},
+	}
+	return diskSpec
+}
+
+func (c *gceClient) getInstanceNetworkSpec(privateIPOnly bool) []*compute.NetworkInterface {
+	networkURL := c.getNetworkURL()
+	subNetworkURL := c.getSubNetworkURL()
+
+	accessConfig := &compute.AccessConfig{
+		Name:        "External NAT",
+		NetworkTier: "STANDARD",
```

Comment:
[Not a blocker, let's add an issue for later?] I think we should make this configurable in provider.yaml, and even allow overriding it via an annotation. The price difference is not that significant: https://cloud.google.com/network-tiers/pricing but the premium tier sounds good and desirable for certain customers.

Reply:
Yes, I agree there are a few of these I want to add as annotations, another one is the `SATA` vs `NVMe` for disk interface type (line 98 instances.go)